lib/IDE/CompletionInstance.cpp

```diff
@@ -0,0 +1,360 @@
+/// cannot reuse the CompilerInstance.
+static llvm::hash_code calculateInvocationHash(const CompilerInvocation &Inv) {
+  llvm::hash_code hash(0);
+
+  auto &frontendOpts = Inv.getFrontendOptions();
+  for (const InputFile &Input : frontendOpts.InputsAndOutputs.getAllInputs())
+    hash = llvm::hash_combine(hash, Input.file());
+  hash = llvm::hash_combine(
+      hash,
+      frontendOpts.InputKind,
+      llvm::makeArrayRef(frontendOpts.ImplicitImportModuleNames),
+      frontendOpts.ImplicitObjCHeaderPath,
+      frontendOpts.ModuleName,
+      frontendOpts.PrebuiltModuleCachePath,
+      llvm::makeArrayRef(frontendOpts.PreferInterfaceForModules),
+      frontendOpts.ParseStdlib,
+      frontendOpts.EnableSourceImport,
+      frontendOpts.ImportUnderlyingModule);
+
+  auto &langOpts = Inv.getLangOptions();
+  hash = llvm::hash_combine(
+      hash,
+      langOpts.Target.str(),
+      llvm::VersionTuple(langOpts.EffectiveLanguageVersion).getAsString(),
+      llvm::VersionTuple(langOpts.PackageDescriptionVersion).getAsString(),
+      langOpts.DisableAvailabilityChecking,
+      langOpts.EnableAccessControl,
+      langOpts.EnableAppExtensionRestrictions,
+      langOpts.RequireExplicitAvailability,
+      langOpts.RequireExplicitAvailabilityTarget,
+      langOpts.EnableObjCInterop,
+      langOpts.EnableCXXInterop,
+      langOpts.InferImportAsMember,
+      langOpts.EnableSwift3ObjCInference);
+
+  auto &searchPathOpts = Inv.getSearchPathOptions();
+  hash = llvm::hash_combine(
+      hash,
+      searchPathOpts.SDKPath,
+      llvm::makeArrayRef(searchPathOpts.ImportSearchPaths),
+      llvm::makeArrayRef(searchPathOpts.VFSOverlayFiles),
+      searchPathOpts.RuntimeResourcePath,
+      llvm::makeArrayRef(searchPathOpts.RuntimeLibraryImportPaths),
+      llvm::makeArrayRef(searchPathOpts.SkipRuntimeLibraryImportPaths));
+  for (auto &P : searchPathOpts.FrameworkSearchPaths)
+    hash = llvm::hash_combine(hash, P.IsSystem, P.Path);
+
+  auto &clangOpts = Inv.getClangImporterOptions();
+  hash = llvm::hash_combine(
+      hash,
+      clangOpts.ModuleCachePath,
+      llvm::makeArrayRef(clangOpts.ExtraArgs),
+      clangOpts.OverrideResourceDir,
+      clangOpts.TargetCPU,
+      static_cast<uint8_t>(clangOpts.Mode),
+      clangOpts.ImportForwardDeclarations,
+      clangOpts.InferImportAsMember,
+      clangOpts.DisableSwiftBridgeAttr,
+      clangOpts.DisableOverlayModules);
+
+  return hash;
+}
+
+} // namespace
+
+CompilerInstance *CompletionInstance::getReusingCompilerInstance(
+    const swift::CompilerInvocation &Invocation,
+    llvm::MemoryBuffer *completionBuffer, unsigned int Offset,
+    DiagnosticConsumer *DiagC) {
+  if (!EnableASTCaching)
+    return nullptr;
+
+  if (CurrentASTReuseCount >= MaxASTReuseCount) {
+    CurrentASTReuseCount = 0;
+    return nullptr;
+  }
+
+  if (!CachedCI)
+    return nullptr;
+
+  if (calculateInvocationHash(Invocation) !=
+      calculateInvocationHash(CachedCI->getInvocation()))
```

Comment:
Why do we support non-identical compiler invocations? When does that happen in practice that we care about?

Reply:
> Why do we support non-identical compiler invocations?

What do you mean? Are you suggesting to implement `CompilerInvocation::operator==` unless support non-identical compiler invocations?

Reply:
If we only care about equality, could we compare the compiler arguments directly as strings instead? What I'm concerned about is enumerating the options in this file is hard to keep up to date when we add new options.

Reply:
I see. I completely agree with your concern.
Just comparing `Args` was what I implemented at first. The reason I stop doing that was I just didn't want to pass them to this place. But I will reconsider that.

Reply:
Modified to remember the hash of the arguments of the invocation.